lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -0,0 +1,580 @@
+              if (p.value() == node)
+                portNum = p.index();
+            }
+            auto remark = ins.emitRemark("instance is part of a combinational "
+                                         "cycle, instance port number '")
+                          << childPortNum << "' has a path from port number '"
+                          << portNum << "'";
+            FieldRef fOut = getFieldRef(node);
+            FieldRef fIn = getFieldRef(childNode.first);
+            auto inName = getFieldName(fIn).first;
+            auto outName = getFieldName(fOut).first;
+            remark << ", " << inName << " <- " << outName;
+          })
+          .Case<MemOp>([&](MemOp mem) {
+            mem.emitRemark("memory is part of a combinational cycle ");
+          })
+          .Case<SubfieldOp, SubindexOp, SubaccessOp>([&](auto op) {})
+          .Default([&](auto owner) {
+            owner->emitRemark(
+                "this operation is part of the combinational cycle");
+          });
+  }
+
+  FModuleOp module;
+  InstanceGraph &instanceGraph;
+  // Set of visiting and visited nodes.
+  llvm::SmallDenseSet<Value> visiting, visited;
+  // Combinational paths between the ports of a module.
+  llvm::SmallDenseMap<Value, llvm::SmallDenseMap<size_t, SmallVector<Node>>>
+      &portPaths;
+  // Map of each unique storage location id and the list of SSA values that
+  // refer to it.
+  llvm::DenseMap<size_t, SmallVector<Value>> storageLocIdToValuesMap;
+  // Map of an SSA value to the list of storage location ids, that it can refer
+  // to.
+  llvm::DenseMap<Value, SmallVector<size_t>> valToStorageLocIdMap;
+  // For each leaf value, record the aggregate base value that it corresponds
+  // to.
+  llvm::DenseMap<Value, Value> leafToBaseValMap;
+  size_t storageLocId = 0;
+  SmallVector<Value> currentPath;
+  bool printCycle = true;
+  bool detectedCycle = false;
+};
+
+/// This pass constructs a local graph for each module to detect combinational
+/// cycles. To capture the cross-module combinational cycles, this pass inlines
+/// the combinational paths between IOs of its subinstances into a subgraph and
+/// encodes them in a `combPathsMap`.
+class CheckCombLoopsPass : public CheckCombLoopsBase<CheckCombLoopsPass> {
+public:
+  void runOnOperation() override {
+    auto &instanceGraph = getAnalysis<InstanceGraph>();
+    llvm::SmallDenseMap<Value, llvm::SmallDenseMap<size_t, SmallVector<Node>>>
+        portPaths;
+    // Traverse modules in a post order to make sure the combinational paths
+    // between IOs of a module have been detected and recorded in `combPathsMap`
+    // before we handle its parent modules.
+    for (auto *igNode : llvm::post_order<InstanceGraph *>(&instanceGraph)) {
+      if (auto module = dyn_cast<FModuleOp>(*igNode->getModule())) {
+        ModuleDFS m(module, instanceGraph, portPaths);
+        if (m.processModule().failed())
+          return signalPassFailure();
+      }
```

Comment:
In the future, we might want to separate graph constructions and cycle detection so that we can parallelize the former.

Reply:
There are a few parallelization opportunities here. I will work on them next.